build-tools/packages/build-cli/src/commands/generate/sourceEntrypoints.ts

```diff
@@ -0,0 +1,410 @@
+		 * [public, `{ relPath: "./lib/entrypoints/public.js", conditions: [], isTypeOnly: false }`]]
+		 */
+		const mapApiTagToExportPath: Map<ApiTag, ExportData> = mapExportPathToApiTag(
+			packageJson,
+			emitDeclarationOnly,
+			this.logger,
+		);
+
+		/**
+		 * Example of `mapApiTagToSourcePath` for a package where rootDir is ./src/entrypoints/
+		 * [["./src/entrypoints/beta.ts", "beta"],
+		 * ["./src/entrypoints/alpha.ts", "alpha"],
+		 * ["./src/entrypoints/public.ts", "public"],
+		 * ["./src/entrypoints/legacy.ts", "legacy"]]
+		 */
+		const mapApiTagToSourcePath: Map<string, ApiTag> = await mapSourcePathToApiTag(rootDir);
+
+		/**
+		 * Example of `mapSourceToExportPath` for a package
+		 * [[beta, `{ relPath: "./src/entrypoints/beta.js", conditions: [], isTypeOnly: false }`],
+		 * [public, `{ relPath: "./src/entrypoints/public.js", conditions: [], isTypeOnly: false }`]]
+		 */
+		const mapSourceToExportPath: Map<ApiTag, ExportData> = getOutputConfiguration(
+			mapApiTagToSourcePath,
+			mapApiTagToExportPath,
+			rootDir,
+			tsconfigOutDir,
+			this.logger,
+		);
+
+		if (mapSourceToExportPath.size === 0) {
+			throw new Error(
+				`There are no package exports matching requested output entrypoints:\n\t${[
+					...mapApiTagToSourcePath.keys(),
+				].join("\n\t")}`,
+			);
+		}
+
+		// generate source entrypoints under `${outDir}`
+		return generateSourceEntrypoints(mainEntrypoint, mapSourceToExportPath, this.logger);
+	}
+}
+
+// Formats the given output directory path to ensure it starts with `./` and ends with `/`.
+function formatPath(outDir: string): string {
+	return `./${outDir.replace(/^\.\/|\/$/g, "")}/`;
+}
+
+/**
+ * Retrieves `rootDir`, `outDir` and `emitDeclarationOnly` settings from a `tsconfig.json` file.
+ *
+ * @param tsconfigPath - Path to the TypeScript config file.
+ * @returns An object with `rootDir`, `outDir` and `emitDeclarationOnly` values.
+ * @throws If `rootDir` and `outDir` is not defined in the config file.
+ */
+async function getTsConfigCompilerOptions(
+	tsconfigPath: string,
+): Promise<{ rootDir: string; tsconfigOutDir: string; emitDeclarationOnly: boolean }> {
+	const formatTsconfigPath = formatPath(tsconfigPath);
+
+	const tsConfigContent = await fs.readFile(`${formatTsconfigPath}tsconfig.json`, {
+		encoding: "utf8",
+	});
+
+	if (tsConfigContent === undefined) {
+		throw new Error(`tsconfig.json not found in ${formatTsconfigPath}`);
+	}
+
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+	const tsconfig = JSON5.parse(tsConfigContent);
+
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+	const { compilerOptions } = tsconfig;
+
+	if (compilerOptions === undefined) {
+		throw new Error(`No compilerOptions defined in ${formatTsconfigPath}tsconfig.json`);
+	}
+
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+	const { rootDir, outDir, emitDeclarationOnly } = compilerOptions;
```

Comment:
If we need emitDeclarationOnly (alternative would be to retry with "types" when "default" gives us nothing), then we really want tsconfig extends processing. (I commented, I think, in the usage example PR that it is not good to duplicate settings. It is going to get messed up by humans.)
I expect that rootDir and outDir are usually present based on how we usually write things and not in danger.

Reply:
`build-tools/packages/build-cli/src/library/repoPolicyCheck/fluidBuildDatabase.ts`'s `tscOutput` contains all the smarts for accurately deriving tsc options including rootDir and outDir. (You would likely give `tscUtils.parseCommandLine` fake command line `tsc --project <tsconfigPath>` to get results.)

Reply:
Reading `emitDeclarationOnly` simplifies concluding if the pacakge is types only. resolve.exports have options to first check default and then types but that makes reading types only value difficult/messy.